doc/decisions/DR016-One-to-many-result-paging.md

```diff
@@ -0,0 +1,522 @@
+
+### Outcome
+
+Given the reported difficulty of crafting queries that are both batch
+and paging-aware, we opt to go with Option 1 and split the paging API(s)
+from the batched API(s).
+
+This then leaves only two remaining solutions: (non-batched) page token
+or cursor objects.
+
+## Considering manager implementations
+
+### Using Python's DB-API standard
+
+#### Batch signature
+```python
+# Callback taking index of batch element and all results for that element.
+SuccessCallback = Callable[[int, List[EntityReference]], None]
+
+def relatedEntities(
+        batch: List[EntityReference], relationship,
+        successCb: SuccessCallback, errorCb):
+
+    cursor = get_my_connection().cursor()
+
+    # Params for query.
+    my_relationship_type: str = get_my_relationship_type(relationship)
+    entity_ref_strs = (str(ref) for ref in batch)
+
+    # Run the query.
+    placeholders = ", ".join(["?"] * len(batch))  # "?, ?, ?, ..."
+    cursor.execute(
+        f"""
+        SELECT FromRef, ToRef FROM Relationship
+        WHERE RelationshipType = ?
+        AND FromRef IN ({placeholders})
+        ORDER BY FromRef
+        """, my_relationship_type, *entity_ref_strs)
+
+    # Get the raw results
+    result_set = cursor.fetchall()
+
+    # Gather the results for each input ref.
+    result_dict = collections.defaultdict(list)
+    for result in result_set:
+        from_ref = EntityReference(result[0])
+        to_ref = EntityReference(result[1])
+        result_dict[from_ref].append(to_ref)
+
+    # Call callbacks.
+    for idx, from_ref in enumerate(batch):
+        successCb(idx, result_dict[from_ref])
+```
+
+#### Page token signature
+
+```python
+class PageToken(AbstractPageToken):
+    def __init__(self, cursor):
+        self.cursor = cursor
+
+
+def relatedEntities(
+    input: EntityReference, relationship, limit=-1, pageToken=None
+) -> Tuple[AbstractPageToken, List[EntityReferences]]:
+
+    if pageToken is None:
+        my_relationship_type: str = get_my_relationship_type(relationship)
+
+        cursor = get_my_connection().cursor()
+        cursor.execute(
+            f"""
+            SELECT ToRef FROM Relationship
+            WHERE RelationshipType = ?
+            AND FromRef = ?
+            """, my_relationship_type, str(input))
+
+        pageToken = PageToken(cursor)
+
+    if limit <= 0:
+        result_set = pageToken.cursor.fetchall()
+    else:
+        result_set = pageToken.cursor.fetchmany(limit)
+
+    related_refs = [EntityReference(result[0]) for result in result_set]
+    return pageToken, related_refs
+```
+
+#### Cursor object signature
+
+The design of DB-API cursors have a somewhat different control flow to
+the `AbstractPage` example used throughout, and in practice we would
+probably like to mirror a standard such as DB-API in any final design.
+However, shimming the DB-API cursor to match our contrived and
+ill-conceived `AbstractPage` is a useful exercise.
+
+```python
+class Page(AbstractPage):
```

Comment:
[fix] Can we align the limit behaviour between the two implementations? As iirir, this version doesn't have a way to define the size of the first page.

ie. pass limit in `relatedEntities` and set the initial page size there?

Reply:
The `next` function determines the page size applied to the next `get` call.

Reply:
But if I just call `get` on the cursor im given by the method, it gets everything defeating paging.

Reply:
In the current wonky design the initial page size is 0, so it'll be `fetchmany(0)` and you'll get nothing.

Having said that, I agree with your assessment in https://github.com/OpenAssetIO/OpenAssetIO/pull/908#discussion_r1198864935 and will redesign the wonky `AbstractPage` to make a bit more sense.

Reply:
🙏